File: clients/typescript/examples/debugger.ts
import { BashClient, BreakHitEvent } from "../src";

async function main() {
  const socketPath = process.argv[2] ?? "/tmp/bash-server-1000/sock";
  const token = process.argv[3] ?? "a".repeat(64);

  const client = await BashClient.connect(socketPath);
  try {
    await client.auth(token);
    await client.debug.enable();

    const bpId = await client.debug.addBreakpoint({ kind: "command", pattern: "echo" });
    console.log(`Breakpoint set: id=${bpId}`);

    client.debug.on("break_hit", async (e: BreakHitEvent) => {
      console.log(`[BREAK] line=${e.line} cmd=${e.command}`);
      const ast = await client.debug.inspectAst();
      console.log(`  AST:`, JSON.stringify(ast));
      await client.debug.continue_();
    });

    const result = await client.eval("echo breakpoint_test");
    console.log(`Result: ${result.stdout}`);

    await client.debug.removeBreakpoint(bpId);
    await client.debug.disable();
  } finally {
    await client.close();
  }
}

main().catch(console.error);

File: clients/python/examples/debugger.py
#!/usr/bin/env python3
"""Example: set breakpoints, step through commands, inspect AST."""

import asyncio
import sys
from bashclient import BashClient, BreakHitEvent


async def main():
    socket_path = sys.argv[1] if len(sys.argv) > 1 else "/tmp/bash-server/sock"
    token = sys.argv[2] if len(sys.argv) > 2 else "a" * 64

    async with await BashClient.connect(socket_path) as client:
        await client.auth(token)

        # Enable debugger
        await client.debug.enable()

        # Set a breakpoint on 'echo' commands
        bp_id = await client.debug.add_breakpoint(kind="command", pattern="echo")
        print(f"Breakpoint set: id={bp_id}")

        # List breakpoints
        breakpoints = await client.debug.list_breakpoints()
        for bp in breakpoints:
            print(f"  BP#{bp.id}: kind={bp.kind} enabled={bp.enabled}")

        # Register break handler
        async def on_break(event: BreakHitEvent):
            print(f"[BREAK] line={event.line} cmd={event.command!r}")
            ast = await client.debug.inspect_ast()
            print(f"  AST: {ast}")
            await client.debug.continue_()

        client.debug.on("break_hit", on_break)

        # Run a command that will hit the breakpoint
        result = await client.eval("echo breakpoint_test")
        print(f"Result: {result.stdout}")

        # Cleanup
        await client.debug.remove_breakpoint(bp_id)
        await client.debug.disable()


if __name__ == "__main__":
    asyncio.run(main())
